### PR TITLE
Migrate linter and spec from coffeescript to javascript

### PR DESCRIPTION
**Pull request by andystanton**

### CODE REVIEW

File: lib/linter-scalac.js [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    },
+    compileAllClassesOnLint: {
+      type: 'boolean',
+      default: false,
+      order: 4
+    }
+  },
+
+  activate: () => {
+    require('atom-package-deps').install();
```
Comment (Arcanemagus) [inline lib/linter-scalac.js]: Might as well fill in the linter name while you are at it, unless you want to keep this PR a "pure" migration.

File: lib/linter-scalac.js [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+    const CompositeDisposable = require('atom').CompositeDisposable;
```
Comment (Arcanemagus) [inline lib/linter-scalac.js]: This should already be defined from the `import` on line 3.

File: lib/linter-scalac.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    const regex = '(?<file>.+):(?<line>\\d+): (?<type>(error|warning)): (?<message>(.+))';
+    const compilerRegex = 'scalac (?<type>(error|warning)): (?<message>(.+))';
+
+    return {
+      grammarScopes: ['source.scala'],
+      scope: this.compileAllClassesOnLint ? 'project' : 'file',
+      lintOnFly: false,
+      lint: (textEditor) => {
+        const filePath = textEditor.getPath();
+        const command = atom.config.get('linter-scalac.scalacExecutablePath');
```
Comment (Arcanemagus) [inline lib/linter-scalac.js]: This should be moved to an `observe` in `activate()`, again unless you want a "pure" migration.

File: lib/linter-scalac.js (hunk 1 above)
Comment (andystanton) [inline lib/linter-scalac.js]: Not fussed about the purity of the migration. What format should the name take? Is it "linter-scalac" or just "scalac"?

File: lib/linter-scalac.js [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        const matchingProjectPaths = atom.project.getPaths().filter(tmpPath =>
```
Comment (Arcanemagus) [inline lib/linter-scalac.js]: This whole project path section can mostly be replaced with a call to `atom.project.relativizePath(filePath)`.

File: lib/linter-scalac.js (hunk 1 above)
Comment (Arcanemagus) [inline lib/linter-scalac.js]: Just "scalac", it's meant to identify the linter in messages so if a user has multiple linters running on a file they can tell which one is sending which message.

File: spec/linter-scalac-spec.js [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+
+  const exec = require('promised-exec');
+  const mkdirp = require('mkdirp');
+  const rm = require('rimraf');
+
+  const lint = require('../lib/linter-scalac')
+    .provideLinter()
+    .lint;
+
+  const fixturesPath = `${__dirname}${path.sep}fixtures`;
```
Comment (Arcanemagus) [inline spec/linter-scalac-spec.js]: Should just use `path.join`, no need for the templatized strings or using `path.sep`.

Comment (Arcanemagus) [inline spec/linter-scalac-spec.js]: Looks like this comment applies to most of the file. :stuck_out_tongue:

**Comment by Arcanemagus** [comment]
Feel free to merge whenever, the specs are working as is, using `path.join()` would just simplify things quite a bit.

**Comment by andystanton** [comment]
Thanks for the review @Arcanemagus. I've made those changes and squashed the commits. Any more comments before I merge?

**Comment by Arcanemagus** [comment]
Soooo _many_ paths lol, anyway this looks good to me :wink:.

**Pull request merged**